moz_security/sandboxes/heka/output/moz_security_tigerblood.lua

```diff
@@ -0,0 +1,270 @@
+        ts = artifacts.ts or os.time(),
+        nonce = artifacts.nonce or nonce(6), -- six b64 chars
+        app = artifacts.app,
+        dlg = artifacts.dlg,
+    }
+
+    if options.ext then
+        mac_artifacts.ext = options.ext
+    end
+
+    if options.payload then
+        mac_artifacts.hash = hash_payload(options.payload, credentials.algorithm, options.content_type)
+    end
+
+    local mac = calculate_mac(credentials, mac_artifacts)
+
+    local header = string.format("Hawk id=\"%s\", ts=\"%s\", nonce=\"%s\"",  credentials.id, mac_artifacts.ts, mac_artifacts.nonce)
+
+    if mac_artifacts.hash then
+        header = string.format("%s, hash=\"%s\"", header, mac_artifacts.hash)
+    end
+    if options.ext then
+        header = string.format("%s, ext=\"%s\"", header, mac_artifacts.ext)
+    end
+
+    header = string.format("%s, mac=\"%s\"", header, mac)
+
+    return header
+end
+
+-- IP Reputation Output
+
+local http = require("socket.http")
+local https = require("ssl.https")
+local url = require("socket.url")
+local ltn12 = require("ltn12")
+local cjson = require("cjson")
+
+local client_config = { -- config params set via configure
+    base_url = nil,
+    id = nil,
+    _key = nil,
+    algorithm = "sha256",
+    requestor = nil,
+}
+
+local computed_client_config = { -- config params computed from config
+    host = nil,
+    post = nil,
+}
+
+-- private function for parsing the port out of a URL
+local function get_port(parsed_url)
+    if parsed_url["port"] then
+        return parsed_url["port"]
+    elseif parsed_url["scheme"] and parsed_url["scheme"] == "http" then
+        return 80
+    elseif parsed_url["scheme"] and parsed_url["scheme"] == "https" then
+        return 443
+    else
+        error("No scheme or port found for base_url.")
+    end
+end
+
+-- sets client config values; must be called before making requests
+-- example usage:
+--
+-- configure_client({
+--    base_url = "http://localhost:8080", -- tigerblood service url w/o slash
+--    id = "root", -- hawk ID
+--    _key = "toor", -- hawk key
+-- })
+--
+local function configure_client(new_client_config)
+    client_config["base_url"] = new_client_config["base_url"];
+    client_config["id"] = new_client_config["id"];
+    client_config["_key"] = new_client_config["_key"];
+    client_config["algorithm"] = new_client_config["algorithm"] or "sha256";
+
+    local parsed_base = url.parse(new_client_config["base_url"])
+    computed_client_config["host"] = parsed_base["host"]
+    computed_client_config["port"] = get_port(parsed_base)
+
+    if parsed_base["scheme"] == "https" then
+        client_config["requestor"] = https
+    else
+        client_config["requestor"] = http
+    end
+end
+
+
+-- private function for making hawk-signed HTTP requests
+local function json_request(method, uri, body)
+    local body_json = ""
+    if body then
+        body_json = body
+    end
+
+    local headers = {
+        Authorization = hawk_header(
```

Comment:
Args 2,3,4 can be constructed once outside of this function.  The Arg 3 table can be updated before the call.